modules/ndbm/sdbm/sdbm.3
.\" orig-Id: sdbm.3,v 1.2 90/12/13 13:00:57 oz Exp
.\" $Id$
.TH SDBM 3 "1 March 1990"
.SH NAME
sdbm, dbm_open, dbm_prep, dbm_close, dbm_fetch, dbm_store, dbm_delete, dbm_firstkey, dbm_nextkey, dbm_hash, dbm_rdonly, dbm_error, dbm_clearerr, dbm_dirfno, dbm_pagfno \- data base subroutines
.SH SYNOPSIS
.nf
.ft B
#include <sdbm.h>
.sp
typedef struct {
	char *dptr;
	int dsize;
} datum;
.sp
datum nullitem = { NULL, 0 };
.sp
\s-1DBM\s0 *dbm_open(char *file, int flags, int mode)
.sp
\s-1DBM\s0 *dbm_prep(char *dirname, char *pagname, int flags, int mode)
.sp
void dbm_close(\s-1DBM\s0 *db)
.sp
datum dbm_fetch(\s-1DBM\s0 *db, key)
.sp
int dbm_store(\s-1DBM\s0 *db, datum key, datum val, int flags)
.sp
int dbm_delete(\s-1DBM\s0 *db, datum key)
.sp
datum dbm_firstkey(\s-1DBM\s0 *db)
.sp
datum dbm_nextkey(\s-1DBM\s0 *db)
.sp
long dbm_hash(char *string, int len)
.sp
int dbm_rdonly(\s-1DBM\s0 *db)
int dbm_error(\s-1DBM\s0 *db)
dbm_clearerr(\s-1DBM\s0 *db)
int dbm_dirfno(\s-1DBM\s0 *db)
int dbm_pagfno(\s-1DBM\s0 *db)
.ft R
.fi
.SH DESCRIPTION
.IX "database library" sdbm "" "\fLsdbm\fR"
.IX dbm_open "" "\fLdbm_open\fR \(em open \fLsdbm\fR database"
.IX dbm_prep "" "\fLdbm_prep\fR \(em prepare \fLsdbm\fR database"
.IX dbm_close "" "\fLdbm_close\fR \(em close \fLsdbm\fR routine"
.IX dbm_fetch "" "\fLdbm_fetch\fR \(em fetch \fLsdbm\fR database data"
.IX dbm_store "" "\fLdbm_store\fR \(em add data to \fLsdbm\fR database"
.IX dbm_delete "" "\fLdbm_delete\fR \(em remove data from \fLsdbm\fR database"
.IX dbm_firstkey "" "\fLdbm_firstkey\fR \(em access \fLsdbm\fR database"
.IX dbm_nextkey "" "\fLdbm_nextkey\fR \(em access \fLsdbm\fR database"
.IX dbm_hash "" "\fLdbm_hash\fR \(em string hash for \fLsdbm\fR database"
.IX dbm_rdonly "" "\fLdbm_rdonly\fR \(em return \fLsdbm\fR database read-only mode"
.IX dbm_error "" "\fLdbm_error\fR \(em return \fLsdbm\fR database error condition"
.IX dbm_clearerr "" "\fLdbm_clearerr\fR \(em clear \fLsdbm\fR database error condition"
.IX dbm_dirfno "" "\fLdbm_dirfno\fR \(em return \fLsdbm\fR database bitmap file descriptor"
.IX dbm_pagfno "" "\fLdbm_pagfno\fR \(em return \fLsdbm\fR database data file descriptor"
.IX "database functions \(em \fLsdbm\fR"  dbm_open  ""  \fLdbm_open\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_prep  ""  \fLdbm_prep\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_close  ""  \fLdbm_close\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_fetch  ""  \fLdbm_fetch\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_store  ""  \fLdbm_store\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_delete  ""  \fLdbm_delete\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_firstkey  ""  \fLdbm_firstkey\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_nextkey  ""  \fLdbm_nextkey\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_rdonly  ""  \fLdbm_rdonly\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_error  ""  \fLdbm_error\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_clearerr  ""  \fLdbm_clearerr\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_dirfno  ""  \fLdbm_dirfno\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_pagfno  ""  \fLdbm_pagfno\fP
.LP
This package allows an application to maintain a mapping of <key,value> pairs
in disk files.  This is not to be considered a real database system, but is
still useful in many simple applications built around fast retrieval of a data
value from a key.  This implementation uses an external hashing scheme,
called Dynamic Hashing, as described by Per-Aake Larson in BIT 18 (1978) pp.
184-201.  Retrieval of any item usually requires a single disk access.
The application interface is compatible with the
.IR ndbm (3)
library.
.LP
An
.B sdbm
database is kept in two files usually given the extensions
.B \.dir
and
.BR \.pag .
The
.B \.dir
file contains a bitmap representing a forest of binary hash trees, the leaves
of which indicate data pages in the
.B \.pag
file.
.LP
The application interface uses the
.B datum
structure to describe both
.I keys
and
.IR value s.
A
.B datum
specifies a byte sequence of
.I dsize
size pointed to by
.IR dptr .
If you use
.SM ASCII
strings as
.IR key s
or
.IR value s,
then you must decide whether or not to include the terminating
.SM NUL
byte which sometimes defines strings.  Including it will require larger
database files, but it will be possible to get sensible output from a
.IR strings (1)
command applied to the data file.
.LP
In order to allow a process using this package to manipulate multiple
databases, the applications interface always requires a
.IR handle ,
a
.BR "DBM *" ,
to identify the database to be manipulated.  Such a handle can be obtained
from the only routines that do not require it, namely
.BR dbm_open (\|)
or
.BR dbm_prep (\|).
Either of these will open or create the two necessary files.  The
difference is that the latter allows explicitly naming the bitmap and data
files whereas
.BR dbm_open (\|)
will take a base file name and call
.BR dbm_prep (\|)
with the default extensions.
The
.I flags
and
.I mode
parameters are the same as for
.BR open (2).
.LP
To free the resources occupied while a database handle is active, call
.BR dbm_close (\|).
.LP
Given a handle, one can retrieve data associated with a key by using the
.BR dbm_fetch (\|)
routine, and associate data with a key by using the
.BR dbm_store (\|)
routine.
.LP
The values of the
.I flags
parameter for
.BR dbm_store (\|)
can be either
.BR \s-1DBM_INSERT\s0 ,
which will not change an existing entry with the same key, or
.BR \s-1DBM_REPLACE\s0 ,
which will replace an existing entry with the same key.
Keys are unique within the database.
.LP
To delete a key and its associated value use the
.BR dbm_delete (\|)
routine.
.LP
To retrieve every key in the database, use a loop like:
.sp
.nf
.ft B
for (key = dbm_firstkey(db); key.dptr != NULL; key = dbm_nextkey(db))
        ;
.ft R
.fi
.LP
The order of retrieval is unspecified.
.LP
If you determine that the performance of the database is inadequate or
you notice clustering or other effects that may be due to the hashing
algorithm used by this package, you can override it by supplying your
own
.BR dbm_hash (\|)
routine.  Doing so will make the database unintelligable to any other
applications that do not use your specialized hash function.
.sp
.LP
The following macros are defined in the header file:
.IP
.BR dbm_rdonly (\|)
returns true if the database has been opened read\-only.
.IP
.BR dbm_error (\|)
returns true if an I/O error has occurred.
.IP
.BR dbm_clearerr (\|)
allows you to clear the error flag if you think you know what the error
was and insist on ignoring it.
.IP
.BR dbm_dirfno (\|)
returns the file descriptor associated with the bitmap file.
.IP
.BR dbm_pagfno (\|)
returns the file descriptor associated with the data file.
.SH SEE ALSO
.IR open (2).
.SH DIAGNOSTICS
Functions that return a
.B "DBM *"
handle will use
.SM NULL
to indicate an error.
Functions that return an
.B int
will use \-1 to indicate an error.  The normal return value in that case is 0.
Functions that return a
.B datum
will return
.B nullitem
to indicate an error.
.LP
As a special case of
.BR dbm_store (\|),
if it is called with the
.B \s-1DBM_INSERT\s0
flag and the key already exists in the database, the return value will be 1.
.LP
In general, if a function parameter is invalid,
.B errno
will be set to
.BR \s-1EINVAL\s0 .
If a write operation is requested on a read-only database,
.B errno
will be set to
.BR \s-1ENOPERM\s0 .
If a memory allocation (using
.IR malloc (3))
failed,
.B errno
will be set to
.BR \s-1ENOMEM\s0 .
For I/O operation failures
.B errno
will contain the value set by the relevant failed system call, either
.IR read (2),
.IR write (2),
or
.IR lseek (2).
.SH AUTHOR
.IP "Ozan S. Yigit" (oz@nexus.yorku.ca)
.SH BUGS
The sum of key and value data sizes must not exceed
.B \s-1PAIRMAX\s0
(1008 bytes).
.LP
The sum of the key and value data sizes where several keys hash to the
same value must fit within one bitmap page.
.LP
The
.B \.pag
file will contain holes, so its apparent size is larger than its contents.
When copied through the filesystem the holes will be filled.
.LP
The contents of
.B datum
values returned are in volatile storage.  If you want to retain the values
pointed to, you must copy them immediately before another call to this package.
.LP
The only safe way for multiple processes to (read and) update a database at
the same time, is to implement a private locking scheme outside this package
and open and close the database between lock acquisitions.  It is safe for
multiple processes to concurrently access a database read-only.
.SH APPLICATIONS PORTABILITY
For complete source code compatibility with the Berkeley Unix
.IR ndbm (3)
library, the 
.B sdbm.h
header file should be installed in
.BR /usr/include/ndbm.h .
.LP
The
.B nullitem
data item, and the
.BR dbm_prep (\|),
.BR dbm_hash (\|),
.BR dbm_rdonly (\|),
.BR dbm_dirfno (\|),
and
.BR dbm_pagfno (\|)
functions are unique to this package.


CHANGES
$Id$

For detailed revision history see History file

Changes since 1.0 release;
=========================

	Added new (experimental) builtin function groups:
		"DBM" -- "ndbm" keyed database access
		"COM" -- COM/OLE Automation on Win32
		"STCL" -- SNOBOL/Tcl interface (optional)
	Add HTTP.POST function to http.sno
	Handle environments where pointers can have high order bit set
	Avoid broken "install" utilities
	autoconf compatibility: Honor CC, CFLAGS, CPPFLAGS,
		LDFLAGS, DESTDIR environment variables.
	Fix compilation on pre-panther OSX
	Fix second argument to DEFINE(): was not case folded!
	Fix backtracking over unevaluated expressions
	Prepend &FILE and &LINE to TRACE output
	GC stats & tracing (-g command line option, &GTRACE keyword)

Changes since 0.99.44 release;
=============================
	Added snolib(3) man page for SNOBOL4 library routines
	Treat all bytes with 8th bit set as ALPHA on ASCII systems
		allows UTF-8 encoded characters, and national
		character sets to used in labels and identifiers!
	command line:
		-S option to enlarge interpreter stack
	configure:
		added --mandir & --bindir
	IPv6 Support (both client and server)
	bi-directional "pipe" I/O on systems with sockets
	"privileged" port binding for UDP
	Lots of cleanup!
		DJGPP (DOS gcc) fixes
		default to -O3 with gcc
		add C prototypes for all functions
			(function declarations still "K&R" style)
		MINGW (Win32 gcc) fixes
		Fix LOAD() on Darwin (MacOS X)
		Fix timing script on Debian
	Ported to HP-UX/IA64
	VMS:	LOAD() support (not tested)
		Compilation on VMS7.3
		support magic pathnames "/dev/null" and "/dev/tty"
	Win32:	native (low overhead) support for character console I/O
		support magic pathnames "/dev/null" and "/dev/tty"
		fixed UDP I/O
	Fix for parallel make (GNU make, BSD make)
	Compilation under MS Visual Studio .NET 2003
		Make struct descr packable (BITFIELDS_SAME_TYPE)
	Check for oversize integer constants (ERANGE from strtol())

Changes since 0.99.4 release;
============================

	Command line:
		-d (dynamic region size) and -M (pattern match stack size)
			now take values in descriptors
		-v shows CSNOBOL4 version

	Increased default dynamic space to 64K descriptors

	Fix faulty lexical comparison function optimzations for null string
		
	Performance improvments;
		New string hash function
		(reduced both compilation and run time of genc.sno by 5%)

		Optimized most common case (CONTIN actions) in
		"stream" operation used for lexical analysis.
		(reduced compilation time of genc.sno by 8%)

		Merged all scanner (pattern matching) functions into
		single C function (eliminates mutual recursion, stack
		overflows)
			genc runtime reduced 15%
			atn.sno runtime reduced by 63%!! (2.7x faster)
			worst case (4096 char string) reduced 69% (4.8x faster)
			snocone self-compile runtime reduced 5%

		Made pattern building functions (lib/pat.c) inlinable
			(8% reduction in genc.sno compile time)

		Eliminate space padding on lines read by compiler
			(reduced genc.sno compile time by 65%)

		Raised default C compiler optimization levels

	Default &MAXLNGTH to largest possible string

	Documentation:
		README file describes available includes

		Include PDF of manual page

	Implement fatal "Output error" for output, ENDFILE() errors

	Configuration script re-write;
		Eliminate static config files & autoconf script
		configure generates config.m4 and config.h
		(with all defines) for use by loadable functions

	New configuration options;
	**** see "INSTALL" file for more information ****
	--add-define=
		Add a #define to config.h (visible to external functions)
	--double
		EXPERIMENTAL option to use C "double" for SNOBOL4 REAL data type
	--fast
		Architecture specific optimizations on native compilers
	--longlong
		EXPERIMENTAL!!
		try to use C "long long" (64-bit integer) for INTEGER datatype
		when longs/pointers are only 32-bits (ILP32 data model).
	--lp64
		Request use of 64-bit long/pointer data model
		if available on this architecture, and not the default.
	--no-opt
		Compile without optimization

	New SNOBOL library files (see README);
		compatibility libraries:
			fence.sno, logic.sno, not.sno
			spitbol.sno, snobol4+.sno
		HOST() function codes:
			host.sno
		URL (http:, ftp:, file:) access functions

	Use large files (64-bit offsets) where available.  Allows access
		and creation of files larger than 2GB

	New functions;
		FUNCTION() -- predicate to test if a function is defined
			** EXPERIMENTAL **
		LABEL() -- predicate to test if a label is defined
			idea from Steve Duff's version of Macro SPITBOL
		ORD() -- inverse of CHAR()
		SERV_LISTEN() -- act as TCP server
		SSET() -- scaled SET() allow large files when INTEGER is 32bits
		VDIFFER() -- returns first arg if two arguments DIFFER()
			idea from Steve Duff's version of Macro SPITBOL

	Extended HOST() functions (see host.sno) for system diagnosis

	HOST() function will accept REAL or integer encoded STRING args

	TIME() function now returns REAL datatype allows both better
		resolution (for small values) and larger range.  Using
		32-bit floats, runtimes larger than four and a half
		hours will lose resolution.

	Statistics reports compiler and execution times as REAL numbers
		Average statement runtime reported in microseconds (us).

	Allow a REAL any place an INTEGER is required (from SNOBOL4+)
		Contexts include TABLE(),  ITEM(),  array  indices,
		INPUT(),  OUTPUT(),  SET(), keyword values, CHAR(),
		RPAD(), LPAD(), FIELD(), COLLECT(), DUMP(), DUPL(),
		OPSYN(), and SUBSTR()

	load.h updates;
		all RETxxx macros safe in any context
		RETSTR() takes single argument, allows NULL pointer
		RETSTR2() has old RETSTR() functionality
		RETINT() and RETREAL() always set return data type.

	All include files needed to build loadable functions
		installed in SNOLIB directory

	Include BSD tsort program; allow builds from scratch
		on platforms with GNU tsort (Linux, Cygwin)

	New ports;
		Ported to Darwin (MacOS X)


lib/sdbm/sdbm.c
/*
 * sdbm - ndbm work-alike hashed database library
 * based on Per-Aake Larson's Dynamic Hashing algorithms. BIT 18 (1978).
 * author: oz@nexus.yorku.ca
 * status: public domain.
 *
 * core routines
 */

/*
 * orig-Id: sdbm.c,v 1.16 90/12/13 13:01:31 oz Exp
 * $Id$
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sdbm.h"
#include "tune.h"
#include "pair.h"

#include <sys/types.h>
#include <sys/stat.h>
#ifdef BSD42
#include <sys/file.h>
#else
#include <fcntl.h>
#include <memory.h>
#endif
#include <errno.h>
#include <string.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>			/* POSIX.1 SEEK_SET */
#else
#include <stdio.h>			/* MSVC SEEK_SET */
#endif

#ifdef __STDC__
#include <stddef.h>
#include <stdlib.h>			/* phil: for malloc */
#endif

#ifndef NULL
#define NULL	0
#endif

/*
 * externals
 */
#ifndef sun
extern int errno;
#endif

#if 0	/* phil: DJGPP complains; all unnecessary? */
extern char *malloc proto((unsigned int));
extern void free proto((void *));
extern long lseek();
#endif

/*
 * forward
 */
static int getdbit proto((DBM *, long));
static int setdbit proto((DBM *, long));
static int getpage proto((DBM *, long));
static datum getnext proto((DBM *));
static int makroom proto((DBM *, long, int));

/*
 * useful macros
 */
#define bad(x)		((x).dptr == NULL || (x).dsize <= 0)
#define exhash(item)	dbm_hash((item).dptr, (item).dsize)
#define ioerr(db)	((db)->flags |= DBM_IOERR)

#define OFF_PAG(off)	(long) (off) * PBLKSIZ
#define OFF_DIR(off)	(long) (off) * DBLKSIZ

static long masks[] = {
	000000000000, 000000000001, 000000000003, 000000000007,
	000000000017, 000000000037, 000000000077, 000000000177,
	000000000377, 000000000777, 000000001777, 000000003777,
	000000007777, 000000017777, 000000037777, 000000077777,
	000000177777, 000000377777, 000000777777, 000001777777,
	000003777777, 000007777777, 000017777777, 000037777777,
	000077777777, 000177777777, 000377777777, 000777777777,
	001777777777, 003777777777, 007777777777, 017777777777
};

datum nullitem = {NULL, 0};

DBM *
dbm_open(file, flags, mode)
register char *file;
register int flags;
register int mode;
{
	register DBM *db;
	register char *dirname;
	register char *pagname;
	register int n;

	if (file == NULL || !*file)
		return errno = EINVAL, (DBM *) NULL;
/*
 * need space for two seperate filenames
 */
	n = strlen(file) * 2 + strlen(DIRFEXT) + strlen(PAGFEXT) + 2;

	if ((dirname = malloc((unsigned) n)) == NULL)
		return errno = ENOMEM, (DBM *) NULL;
/*
 * build the file names
 */
	dirname = strcat(strcpy(dirname, file), DIRFEXT);
	pagname = strcpy(dirname + strlen(dirname) + 1, file);
	pagname = strcat(pagname, PAGFEXT);

	db = dbm_prep(dirname, pagname, flags, mode);
	free((char *) dirname);
	return db;
}

DBM *
dbm_prep(dirname, pagname, flags, mode)
char *dirname;
char *pagname;
int flags;
int mode;
{
	register DBM *db;
	struct stat dstat;

	if ((db = (DBM *) malloc(sizeof(DBM))) == NULL)
		return errno = ENOMEM, (DBM *) NULL;

        db->flags = 0;
        db->hmask = 0;
        db->blkptr = 0;
        db->keyptr = 0;
/*
 * adjust user flags so that WRONLY becomes RDWR, 
 * as required by this package. Also set our internal
 * flag for RDONLY if needed.
 */
	if (flags & O_WRONLY)
		flags = (flags & ~O_WRONLY) | O_RDWR;

	else if ((flags & 03) == O_RDONLY)
		db->flags = DBM_RDONLY;
/*
 * open the files in sequence, and stat the dirfile.
 * If we fail anywhere, undo everything, return NULL.
 */
	if ((db->pagf = open(pagname, flags, mode)) > -1) {
		if ((db->dirf = open(dirname, flags, mode)) > -1) {
/*
 * need the dirfile size to establish max bit number.
 */
			if (fstat(db->dirf, &dstat) == 0) {
/*
 * zero size: either a fresh database, or one with a single,
 * unsplit data page: dirpage is all zeros.
 */
				db->dirbno = (!dstat.st_size) ? 0 : -1;
				db->pagbno = -1;
				db->maxbno = dstat.st_size * BYTESIZ;

				(void) memset(db->pagbuf, 0, PBLKSIZ);
				(void) memset(db->dirbuf, 0, DBLKSIZ);
			/*
			 * success
			 */
				return db;
			}
			(void) close(db->dirf);
		}
		(void) close(db->pagf);
	}
	free((char *) db);
	return (DBM *) NULL;
}

void
dbm_close(db)
register DBM *db;
{
	if (db == NULL)
		errno = EINVAL;
	else {
		(void) close(db->dirf);
		(void) close(db->pagf);
		free((char *) db);
	}
}

datum
dbm_fetch(db, key)
register DBM *db;
datum key;
{
	if (db == NULL || bad(key))
		return errno = EINVAL, nullitem;

	if (getpage(db, exhash(key)))
		return getpair(db->pagbuf, key);

	return ioerr(db), nullitem;
}

int
dbm_delete(db, key)
register DBM *db;
datum key;
{
	if (db == NULL || bad(key))
		return errno = EINVAL, -1;
	if (dbm_rdonly(db))
		return errno = EPERM, -1;

	if (getpage(db, exhash(key))) {
		if (!delpair(db->pagbuf, key))
			return -1;
/*
 * update the page file
 */
		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return ioerr(db), -1;

		return 0;
	}

	return ioerr(db), -1;
}

int
dbm_store(db, key, val, flags)
register DBM *db;
datum key;
datum val;
int flags;
{
	int need;
	register long hash;

	if (db == NULL || bad(key))
		return errno = EINVAL, -1;
	if (dbm_rdonly(db))
		return errno = EPERM, -1;

	need = key.dsize + val.dsize;
/*
 * is the pair too big (or too small) for this database ??
 */
	if (need < 0 || need > PAIRMAX)
		return errno = EINVAL, -1;

	if (getpage(db, (hash = exhash(key)))) {
/*
 * if we need to replace, delete the key/data pair
 * first. If it is not there, ignore.
 */
		if (flags == DBM_REPLACE)
			(void) delpair(db->pagbuf, key);
#ifdef SEEDUPS
		else if (duppair(db->pagbuf, key))
			return 1;
#endif
/*
 * if we do not have enough room, we have to split.
 */
		if (!fitpair(db->pagbuf, need))
			if (!makroom(db, hash, need))
				return ioerr(db), -1;
/*
 * we have enough room or split is successful. insert the key,
 * and update the page file.
 */
		(void) putpair(db->pagbuf, key, val);

		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return ioerr(db), -1;
	/*
	 * success
	 */
		return 0;
	}

	return ioerr(db), -1;
}

/*
 * makroom - make room by splitting the overfull page
 * this routine will attempt to make room for SPLTMAX times before
 * giving up.
 */
static int
makroom(db, hash, need)
register DBM *db;
long hash;
int need;
{
	long newp;
	char twin[PBLKSIZ];
	char *pag = db->pagbuf;
	char *new = twin;
	register int smax = SPLTMAX;

	do {
/*
 * split the current page
 */
		(void) splpage(pag, new, db->hmask + 1);
/*
 * address of the new page
 */
		newp = (hash & db->hmask) | (db->hmask + 1);

/*
 * write delay, read avoidence/cache shuffle:
 * select the page for incoming pair: if key is to go to the new page,
 * write out the previous one, and copy the new one over, thus making
 * it the current page. If not, simply write the new page, and we are
 * still looking at the page of interest. current page is not updated
 * here, as dbm_store will do so, after it inserts the incoming pair.
 */
		if (hash & (db->hmask + 1)) {
			if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
			    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
				return 0;
			db->pagbno = newp;
			(void) memcpy(pag, new, PBLKSIZ);
		}
		else if (lseek(db->pagf, OFF_PAG(newp), SEEK_SET) < 0
			 || write(db->pagf, new, PBLKSIZ) < 0)
			return 0;

		if (!setdbit(db, db->curbit))
			return 0;
/*
 * see if we have enough room now
 */
		if (fitpair(pag, need))
			return 1;
/*
 * try again... update curbit and hmask as getpage would have
 * done. because of our update of the current page, we do not
 * need to read in anything. BUT we have to write the current
 * [deferred] page out, as the window of failure is too great.
 */
		db->curbit = 2 * db->curbit +
			((hash & (db->hmask + 1)) ? 2 : 1);
		db->hmask |= db->hmask + 1;

		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return 0;

	} while (--smax);
/*
 * if we are here, this is real bad news. After SPLTMAX splits,
 * we still cannot fit the key. say goodnight.
 */
#ifdef BADMESS
	(void) write(2, "sdbm: cannot insert after SPLTMAX attempts.\n", 44);
#endif
	return 0;

}

/*
 * the following two routines will break if
 * deletions aren't taken into account. (ndbm bug)
 */
datum
dbm_firstkey(db)
register DBM *db;
{
	if (db == NULL)
		return errno = EINVAL, nullitem;
/*
 * start at page 0
 */
	if (lseek(db->pagf, OFF_PAG(0), SEEK_SET) < 0
	    || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
		return ioerr(db), nullitem;
	db->pagbno = 0;
	db->blkptr = 0;
	db->keyptr = 0;

	return getnext(db);
}

datum
dbm_nextkey(db)
register DBM *db;
{
	if (db == NULL)
		return errno = EINVAL, nullitem;
	return getnext(db);
}

/*
 * all important binary trie traversal
 */
static int
getpage(db, hash)
register DBM *db;
register long hash;
{
	register int hbit;
	register long dbit;
	register long pagb;

	dbit = 0;
	hbit = 0;
	while (dbit < db->maxbno && getdbit(db, dbit))
		dbit = 2 * dbit + ((hash & (1 << hbit++)) ? 2 : 1);

	debug(("dbit: %d...", dbit));

	db->curbit = dbit;
	db->hmask = masks[hbit];

	pagb = hash & db->hmask;
/*
 * see if the block we need is already in memory.
 * note: this lookaside cache has about 10% hit rate.
 */
	if (pagb != db->pagbno) { 
/*
 * note: here, we assume a "hole" is read as 0s.
 * if not, must zero pagbuf first.
 */
		if (lseek(db->pagf, OFF_PAG(pagb), SEEK_SET) < 0
		    || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return 0;
		if (!chkpage(db->pagbuf))
			return 0;
		db->pagbno = pagb;

		debug(("pag read: %d\n", pagb));
	}
	return 1;
}

static int
getdbit(db, dbit)
register DBM *db;
register long dbit;
{
	register long c;
	register long dirb;

	c = dbit / BYTESIZ;
	dirb = c / DBLKSIZ;

	if (dirb != db->dirbno) {
		if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
		    || read(db->dirf, db->dirbuf, DBLKSIZ) < 0)
			return 0;
		db->dirbno = dirb;

		debug(("dir read: %d\n", dirb));
	}

	return db->dirbuf[c % DBLKSIZ] & (1 << dbit % BYTESIZ);
}

static int
setdbit(db, dbit)
register DBM *db;
register long dbit;
{
	register long c;
	register long dirb;

	c = dbit / BYTESIZ;
	dirb = c / DBLKSIZ;

	if (dirb != db->dirbno) {
		if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
		    || read(db->dirf, db->dirbuf, DBLKSIZ) < 0)
			return 0;
		db->dirbno = dirb;

		debug(("dir read: %d\n", dirb));
	}

	db->dirbuf[c % DBLKSIZ] |= (1 << dbit % BYTESIZ);

	if (dbit >= db->maxbno)
		db->maxbno += DBLKSIZ * BYTESIZ;

	if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
	    || write(db->dirf, db->dirbuf, DBLKSIZ) < 0)
		return 0;

	return 1;
}

/*
 * getnext - get the next key in the page, and if done with
 * the page, try the next page in sequence
 */
static datum
getnext(db)
register DBM *db;
{
	datum key;

	for (;;) {
		db->keyptr++;
		key = getnkey(db->pagbuf, db->keyptr);
		if (key.dptr != NULL)
			return key;
/*
 * we either run out, or there is nothing on this page..
 * try the next one... If we lost our position on the
 * file, we will have to seek.
 */
		db->keyptr = 0;
		if (db->pagbno != db->blkptr++)
			if (lseek(db->pagf, OFF_PAG(db->blkptr), SEEK_SET) < 0)
				break;
		db->pagbno = db->blkptr;
		if (read(db->pagf, db->pagbuf, PBLKSIZ) <= 0)
			break;
		if (!chkpage(db->pagbuf))
			break;
	}

	return ioerr(db), nullitem;
}
